backends/bmv2/common/control.h

```diff
@@ -39,18 +48,777 @@ class ControlConverter : public Inspector {
+        return "invalid";
+    }
     /// Return 'true' if the table is 'simple'
     bool handleTableImplementation(const IR::Property* implementation, const IR::Key* key,
                                    Util::JsonObject* table, Util::JsonArray* action_profiles,
-                                   BMV2::SharedActionSelectorCheck& selector_check);
-    Util::IJson* convertIf(const CFG::IfNode* node, cstring prefix);
+                                   BMV2::SharedActionSelectorCheck<arch>*) {
+        if (implementation == nullptr) {
+            table->emplace("type", "simple");
+            return true;
+        }
+
+        if (!implementation->value->is<IR::ExpressionValue>()) {
+            ::error(ErrorType::ERR_EXPECTED, "%1%: expected expression for property", implementation);
+            return false;
+        }
+        auto propv = implementation->value->to<IR::ExpressionValue>();
+
+        bool isSimpleTable = true;
+        Util::JsonObject* action_profile;
+        cstring apname;
+
+        if (propv->expression->is<IR::ConstructorCallExpression>()) {
+            auto cc = P4::ConstructorCall::resolve(
+                propv->expression->to<IR::ConstructorCallExpression>(),
+                ctxt->refMap, ctxt->typeMap);
+            if (!cc->is<P4::ExternConstructorCall>()) {
+                ::error(ErrorType::ERR_EXPECTED,
+                        "%1%: expected extern object for property", implementation);
+                return false;
+            }
+            auto ecc = cc->to<P4::ExternConstructorCall>();
+            auto implementationType = ecc->type;
+            auto arguments = ecc->cce->arguments;
+            apname = implementation->controlPlaneName(ctxt->refMap->newName("action_profile"));
+            action_profile = new Util::JsonObject();
+            action_profiles->append(action_profile);
+            action_profile->emplace("name", apname);
+            action_profile->emplace("id", nextId("action_profiles"));
+            action_profile->emplace_non_null("source_info", propv->expression->sourceInfoJsonObj());
+            // TODO(jafingerhut) - add line/col here?
+            // TBD what about the else if cases below?
+
+            auto add_size = [&action_profile, &arguments](size_t arg_index) {
+                auto size_expr = arguments->at(arg_index)->expression;
+                int size;
+                if (!size_expr->is<IR::Constant>()) {
+                    ::error(ErrorType::ERR_EXPECTED, "%1% must be a constant", size_expr);
+                    size = 0;
+                } else {
+                    size = size_expr->to<IR::Constant>()->asInt();
+                }
+                action_profile->emplace("max_size", size);
+            };
+
+            auto actionSelectorName = Standard::ActionSelectorTraits<arch>::typeName();
+            if (implementationType->name == actionSelectorName) {
```

Comment:
Used to be `implementationType->name == BMV2::TableImplementation::actionSelectorName`